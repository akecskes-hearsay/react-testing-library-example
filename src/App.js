import React from 'react'

function getUser() {
    return Promise.resolve({id: '1', name: 'Admin'});
}

const App = () => {
    const [user, setUser] = React.useState(null);

    const loadUser = async () => {
        const user = await getUser();
        setUser(user);
    };

    React.useEffect(() => {
        loadUser()
    }, []);

    if (user) {
        return (
            <div>
                <p>Signed in as {user.name}</p>
                <button
                    id="button-logout"
                    onClick={() => setUser(null)}
                >
                    Logout
                </button>
            </div>)
    }

    return <div>
        <button
            id="button-login"
            onClick={loadUser}
        >
            Login
        </button>
    </div>;
}

export default App;
