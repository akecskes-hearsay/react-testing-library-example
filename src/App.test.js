import React from 'react';
import {render, screen} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import userEvent from "@testing-library/user-event";
import snapshotDiff from 'snapshot-diff';
import App from './App';

describe('App', () => {
    
    test('renders App component then tests snapshot that succeeds without error', async () => {
        const {container} = render(<App/>);
        expect(screen.queryByText(/Signed in as/)).toBeNull();
        expect(await screen.findByText(/Signed in as/)).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });

    test('renders App component then logs out - with regular snapshots', async () => {
        const {container} = render(<App/>);
        expect(await screen.findByText(/Signed in as/)).toBeInTheDocument();

        expect(container).toMatchSnapshot();

        const logoutButton = screen.getByTestId('button-logout')
        userEvent.click(logoutButton)

        expect(container).toMatchSnapshot();

        const loginButton = screen.getByTestId('button-login')
        userEvent.click(loginButton)

        expect(await screen.findByText(/Signed in as/)).toBeInTheDocument();
        expect(container).toMatchSnapshot();

    });

    test('renders App component then logs out - with snapshot diffs', async () => {
        const {container} = render(<App/>);
        expect(await screen.findByText(/Signed in as/)).toBeInTheDocument();

        const snapshot1 = container.cloneNode(true)
        expect(snapshot1).toMatchSnapshot();

        const logoutButton = screen.getByTestId('button-logout')
        screen.logTestingPlaygroundURL()
        userEvent.click(logoutButton)

        const snapshot2 = container.cloneNode(true)
        expect(snapshotDiff(snapshot1, snapshot2)).toMatchSnapshot();

        const loginButton = screen.getByTestId('button-login')
        userEvent.click(loginButton)

        expect(await screen.findByText(/Signed in as/)).toBeInTheDocument();
        const snapshot3 = container.cloneNode(true)
        expect(snapshotDiff(snapshot2, snapshot3)).toMatchSnapshot();
    });

});
