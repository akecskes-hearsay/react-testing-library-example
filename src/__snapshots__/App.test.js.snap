// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders App component then logs out - with regular snapshots 1`] = `
<div>
  <div>
    <p>
      Signed in as 
      Admin
    </p>
    <button
      id="button-logout"
    >
      Logout
    </button>
  </div>
</div>
`;

exports[`App renders App component then logs out - with regular snapshots 2`] = `
<div>
  <div>
    <button
      id="button-login"
    >
      Login
    </button>
  </div>
</div>
`;

exports[`App renders App component then logs out - with regular snapshots 3`] = `
<div>
  <div>
    <p>
      Signed in as 
      Admin
    </p>
    <button
      id="button-logout"
    >
      Logout
    </button>
  </div>
</div>
`;

exports[`App renders App component then logs out - with snapshot diffs 1`] = `
<div>
  <div>
    <p>
      Signed in as 
      Admin
    </p>
    <button
      id="button-logout"
    >
      Logout
    </button>
  </div>
</div>
`;

exports[`App renders App component then logs out - with snapshot diffs 2`] = `
"Snapshot Diff:
- First value
+ Second value

  <div>
    <div>
-     <p>
-       Signed in as 
-       Admin
-     </p>
      <button
-       id=\\"button-logout\\"
+       id=\\"button-login\\"
      >
-       Logout
+       Login
      </button>
    </div>
  </div>"
`;

exports[`App renders App component then logs out - with snapshot diffs 3`] = `
"Snapshot Diff:
- First value
+ Second value

  <div>
    <div>
+     <p>
+       Signed in as 
+       Admin
+     </p>
      <button
-       id=\\"button-login\\"
+       id=\\"button-logout\\"
      >
-       Login
+       Logout
      </button>
    </div>
  </div>"
`;

exports[`App renders App component then tests snapshot that succeeds without error 1`] = `
<div>
  <div>
    <p>
      Signed in as 
      Admin
    </p>
    <button
      id="button-logout"
    >
      Logout
    </button>
  </div>
</div>
`;
